### PR TITLE
Add support for seq split in Domino

### PR DESCRIPTION
**Pull request by duanhx1037**

### CODE REVIEW

**Comment by GuanhuaWang** [comment]
Hi @duanhx1037 

Really appreciated your quick action on this effort! But right now it is far from ready. Please allow me to change the title to `[Draft] add support for seq split in Domino`. Thank you!

File: deepspeed/runtime/domino/transformer.py [hunk 1]
```diff
@@ -260,21 +260,47 @@ def __init__(self,
...
+                                                    bias=config.add_bias_linear,
+                                                    gather_output=False)
+            self.self_attention_sp = CoreAttention(config, self.layer_number, mpu, self_attn_mask_type)
+            self.dense = RowParallelLinearNoComm(query_projection_size,
+                                                config.hidden_size,
+                                                config=config,
+                                                init_method=config.output_layer_init_method,
+                                                bias=config.add_bias_linear,
+                                                input_is_parallel=True,
+                                                skip_bias_add=True)
```
Comment (GuanhuaWang) [inline deepspeed/runtime/domino/transformer.py:301]: these lines are mostly identical to our shardedAttention module, even the new component `self.self_attention_sp` initialized the same way as original `self.core_attention`. I don't see why we need this duplication.

File: deepspeed/runtime/domino/transformer.py [hunk 2]
```diff
@@ -356,20 +382,73 @@ def forward(self, hidden_states, attention_mask, rotary_pos_emb=None):
         if not self.llama_model:
             rotary_pos_emb = None
 
-        attention_output0, attention_bias0  = \
-            self.self_attention(
-                layernorm_output0,
+        if self.input_split_dim == "seq":
+            # Compute full Q, K, V
+            layernorm_output = torch.concat([layernorm_output0, layernorm_output1], dim=0)
```
Comment (GuanhuaWang) [inline deepspeed/runtime/domino/transformer.py:387]: this is wrong. our original layernorm_output0, layernorm_output1 are split on batch as `dim=1`,  concat on `dim=0` is incorrect.

File: deepspeed/runtime/domino/transformer.py [hunk 3]
```diff
@@ -356,20 +382,73 @@ def forward(self, hidden_states, attention_mask, rotary_pos_emb=None):
...
+
+            handle0 = dist.all_reduce(attention_output0, group=self.mpu.get_tensor_model_parallel_group(), async_op=True)
+
+            # seq 1: core attention
+            context_layer1 = self.self_attention_sp(query_layer[:, :, seq_len//2:, :], key_layer, value_layer, attention_mask[:, :, seq_len//2:, :])
+            # Output. [s, b, h]
+            attention_output1, attention_bias1 = self.dense(context_layer1)
+
+            handle1 = dist.all_reduce(attention_output1, group=self.mpu.get_tensor_model_parallel_group(), async_op=True)
+            handle0.wait()
```
Comment (GuanhuaWang) [inline deepspeed/runtime/domino/transformer.py:433]: these are exactly the same as sharedAttention forward, I don't see why we need these duplication here. 

Also please follow our current code hierarchy, not pull up lower layer module implementation code to upper layer module. 

e.g., if any real change need to make on `sharedAttention`, create a similar module say `XYZAttention`, then here in `DominoTransformerLayer` forward function, we can simply call `XYZAttention` module without duplicating XYZAttention module's every line of code of its forward.

File: deepspeed/runtime/domino/transformer.py (hunk 1 above)
Comment (GuanhuaWang) [inline deepspeed/runtime/domino/transformer.py:301]: this is duplication of shardedAttention module's detail implementation.

**Comment by GuanhuaWang** [comment]
@hwchen2017 please review this pr for seq-parallel implementation of Domino. We can continue collaborate on this!
